### PR TITLE
Implement Unpack

### PR DESCRIPTION
**Pull request by bts**

### CODE REVIEW

File: src/Hazel.hs [hunk 1]
```diff
@@ -236,14 +239,24 @@ infer dirs ctx t = case t of
       _ -> throwStackError (CaseArg:dirs)
         "[infer Case] can't case on non-indices: "
 
-    branchCtxs <- imapM (\i vTm -> check (CaseBranch i:dirs) ctx ty vTm)
+    branchCtxs <- imapM (\i vTm -> check (CaseBranch i:dirs) leftovers ty vTm)
```
Comment (joelburget) [inline src/Hazel.hs:295]: nice find

File: src/Hazel.hs [hunk 2]
```diff
@@ -236,14 +239,24 @@ infer dirs ctx t = case t of
...
 
     return (V.head branchCtxs, ty)
 
+  Unpack (cTm, usage) (vTm, ty) -> do
+    (leftovers, cTmTy) <- infer (Unpack':dirs) ctx cTm
+    tupTys <- case cTmTy of
+                TupleTy allTupTys -> return allTupTys
+                _ -> throwStackError (Unpack':dirs)
+                  ("[infer Unpack] can't unpack a non-tuple: " ++ show cTmTy)
+    let newCtx = foldl' (flip (:)) leftovers ((,usage) <$> tupTys)
```
Comment (joelburget) [inline src/Hazel.hs]: I find this hard to read -- it's a little code-golfey.

Comment (joelburget) [inline src/Hazel.hs]: Maybe break out this `foldl' (flip (:))` pattern into a named function?

**Comment by bts** [comment]
``` patch
diff --git a/src/Hazel.hs b/src/Hazel.hs
index d804d74..df4670b 100644
--- a/src/Hazel.hs
+++ b/src/Hazel.hs
@@ -257,6 +257,9 @@ throwStackError dirs str =
   let stackStr = unlines (map show (reverse dirs))
   in throwError $ stackStr ++ "\n" ++ str

+reversePrependedTo :: Foldable t => t a -> [a] -> [a]
+xs `reversePrependedTo` ys = foldl' (flip (:)) ys xs
+
 infer :: LocationDirections -> Ctx -> Computation -> Either String (Ctx, Type)
 infer dirs ctx t = case t of
   BVar i -> inferVar (BVar':dirs) ctx i
@@ -318,7 +321,7 @@ infer dirs ctx t = case t of
                 TupleTy allTupTys -> return allTupTys
                 _ -> throwStackError (Unpack':dirs)
                   ("[infer Unpack] can't unpack a non-tuple: " ++ show cTmTy)
-    let newCtx = foldl' (flip (:)) leftovers ((,usage) <$> tupTys)
+    let newCtx = ((,usage) <$> tupTys) `reversePrependedTo` leftovers
     leftovers' <- check (Unpack':dirs) newCtx ty vTm
     return (leftovers', ty)

@@ -459,7 +462,7 @@ evalC env tm = case tm of
   Unpack (cTm, _usage) (vTm, _ty) -> do
     cTm' <- evalC env cTm
     case cTm' of
-      Tuple _ tupTms -> evalV (foldl' (flip (:)) env tupTms) vTm
+      Tuple _ tupTms -> evalV (tupTms `reversePrependedTo` env) vTm
       _ -> throwError $ "[evalC Unpack] can't unpack non-tuple: " ++ show cTm'
```

something like this?

**Comment by bts** [comment]
or `reversedAndPrependedTo`?

**Comment by joelburget** [comment]
That looks better. I like `reverseAndPrepend`.

**Comment by cartazio** [comment]
I personally always get uncomfortable with non recursive lets in do
notation, but if that's the style you want in hazel, go for it.  Always
freaks me out though 😭😭😱😱😭

**Comment by bts** [comment]
well, hazel will have a distinction between `let` and `letrec` :)

**Comment by cartazio** [comment]
Ok. That's valid once you get it self hosting. Before then my objection
stands 😭😎

But yet, in a strict language in let normal form, lets are monadic binds so
agreed.  Unlike in Haskell and lazy core irs where case drives everything

**Pull request merged**